homeassistant/components/knx/__init__.py

```diff
@@ -172,6 +180,33 @@ async def async_setup(hass, config):
         schema=SERVICE_KNX_SEND_SCHEMA,
     )
 
+    async def reload_service_handler(service_call: ServiceCallType) -> None:
+        """Remove all KNX components and load new ones from config."""
+
+        # First check for config file. If for some reason it is no longer there
+        # or knx is no longer mentioned, stop the reload.
+        config = await async_integration_yaml_config(hass, DOMAIN)
+
+        if not config or DOMAIN not in config:
+            return
+
+        await hass.data[DOMAIN].xknx.stop()
+
+        registry = await async_get_entity_registry(hass)
+        platforms = async_get_platforms(hass, DOMAIN)
+
+        for platform in platforms:
+            for entity_id, entry in platform.entities.copy().items():
+                await entry.async_remove()
+                if entity_id in registry.entities:
+                    await registry.async_remove(entity_id)
```

Comment:
Don't remove the entity registry entries. That contains all the user customization. Just reset the platforms.
```suggestion
        registry = await async_get_entity_registry(hass)
        platforms = async_get_platforms(hass, DOMAIN)

        for platform in platforms:
            await platform.async_reset()
```

Reply:
Even better to do this using an `asyncio.gather` so it runs in parallel.